azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/toolkit/lib/auth/Account.java

```diff
@@ -95,7 +111,65 @@ public boolean isAvailable() {
     }
 
     public boolean isAuthenticated() {
-        return isAvailable() && this.entity.isAuthenticated();
+        return this.entity.isAuthenticated();
+    }
+
+    protected Mono<Account> authenticateAsync() {
+        return this.verifyAndGetTenantId(createTokenCredential()).map(tenantAndToken -> {
+            if (StringUtils.isBlank(this.entity.getEmail()) && tenantAndToken.getT2() instanceof MsalToken) {
+                String refreshToken = getEmailAndRefreshToken(tenantAndToken.getT2());
+                if (StringUtils.isNotEmpty(refreshToken)) {
+                    AzureEnvironment env = ObjectUtils.firstNonNull(this.entity.getEnvironment(), Azure.az(AzureCloud.class).getOrDefault());
+                    String authority = AzureEnvironmentUtils.getAuthority(env);
+                    this.entity.setTenantCredential(new RefreshTokenTenantCredential(
+                            authority, IdentityConstants.DEVELOPER_SINGLE_SIGN_ON_ID, refreshToken
+                    ));
+                }
+            }
+            if (this.entity.getTenantIds() == null) {
+                this.entity.setTenantIds(tenantAndToken.getT1());
+            }
+            this.entity.setAuthenticated(true);
+            return this;
+        }).flatMap(ignore -> {
+            if (this.entity.getSubscriptions() == null) {
+                return initializeSubscriptions().then(Mono.just(this));
+            }
+            return Mono.just(this);
+        });
+    }
+
+    private String getEmailAndRefreshToken(AccessToken accessToken) {
```

Comment:
Shall we split the method into `refershEmail` and `getRefreshToken`? As now it do return the refresh token, but modify the email address with side effect.

Reply:
there has shared code 
```
        IAuthenticationResult result = ((MsalToken) accessToken).getAuthenticationResult();
        if (result == null) {
            return null;
        }

```

Reply:
Yes, but it is a bit confused that one get method for both email and refresh token